### ISSUE TITLE
Expansion operator results in "Cannot call method 'forEach' of null"

### ISSUE THREAD

**Issue by btiernay** [triAGENS/ArangoDB#1200]
Using ArangoDB 2.3.3, the following AQL query:

```sql
FOR d IN Donor 
  FILTER d.donor_sex == null LIMIT 0, 10
  RETURN { 
    "donorId" : d._donor_id, 
    "genes" : (
      FOR g IN Gene 
        FILTER d._donor_id IN g.donor[*]._donor_id
        RETURN g
    )
  }
```

produces:

```[1503] TypeError: Cannot call method 'forEach' of null (while executing)```

I tried guarding the sub-filter with `g.donor != null && d._donor_id IN g.donor[*]._donor_id` and ` g.donor == null ? false : d._donor_id IN g.donor[*]._donor_id` but this results in the same error.

**Comment by jsteemann** [comment]
Fixed right now for 2.3.5 and 2.4.
There were two issues: 
* the `[*]` operator was pulled out of the original condition, so there was no short-circuit behavior for it. This was necessary in older versions of ArangoDB, but is unnecessary since 2.3. So now the subcondition that includes the `[*]` will be executed only if the previous conditions (e.g. `HAS(...) && IS_LIST(...)` are true. 
* the second issue is that the `[*]` operator assumed that the specified element exists and is an array. This can be considered a feature or an anti-feature. Anyway, the error message in case of no array (`cannot call method 'forEach' of null`) was pretty much useless, so now the error will be prevented.

**Comment by btiernay** [comment]
Thanks!
